### ISSUE TITLE
changing tensor's shape/Keras implementation

### ISSUE THREAD

**Issue by seyedmohammadmortaji** [tobybreckon/fire-detection-cnn#38]
Hi
I am going to implement the model on Keras. In concatenate layer, it doesn't let me to concatenate tensors because of different dimensions. Will flattening them affect the network's performance?

 - OS: Linux
 - Tensorflow Version:  1.14.0
 - OpenCV Version: 4.1.1.26
 
 - Keras version: 2.2.4

**Comment by tobybreckon** [comment]
We have reimplemented the models in Keras successfully beforehand. To clarify which model and which layer are you referring to (FireNet or InceptionV1-OnFire) ?

**Comment by seyedmohammadmortaji** [comment]
Thanks for your help.
The model is inceptionOnFire.
The layer is the first concatenate layer (I think all concatenate layers are the same).
Here is my code:



```
def inc_v1_on_fire(input_shape):
    
    input_layer = K.Input(input_shape)
    conv_7_7 = K.Conv2D(64,
                      7,
                      strides = 2,
                      activation = 'relu',
                      padding = 'same',
                      name = 'conv1_7_7_s2')(input_layer)
    
    pool1_3_3 = K.MaxPooling2D(3,
                            strides = 2,
                            padding = 'same')(conv_7_7)
    pool1_3_3 = LRN()(pool1_3_3)
    
    conv2_3_3_reduce = K.Conv2D(64,
                               1,
                               activation = 'relu',
                               name = 'conv2_3_3_reduce')(pool1_3_3)
    conv2_3_3 = K.Conv2D(128,
                        3,
                        activation='relu',
                        name='conv2_3_3')(conv2_3_3_reduce)
    
    conv2_3_3 = LRN()(conv2_3_3)
    pool2_3_3 = K.MaxPooling2D(3,
                             strides = 2,
                             padding = 'same',
                             name ='pool2_3_3_s2')(conv2_3_3)
    
    inception_3a_1_1 = K.Conv2D(64,
                               1,
                               activation = 'relu',
                               name = 'inception_3a_1_1')(pool2_3_3)
    
    
    inception_3a_3_3_reduce = K.Conv2D(96,
                                      1,
                                      activation ='relu',
                                      name='inception_3a_3_3_reduce')(pool2_3_3)
    
    inception_3a_3_3 = K.Conv2D(128,
                               3,
                               activation = 'relu',
                               name = 'inception_3a_3_3')(inception_3a_3_3_reduce)
    inception_3a_5_5_reduce = K.Conv2D(16,
                                      1,
                                      activation = 'relu',
                                      name ='inception_3a_5_5_reduce')(pool2_3_3)
    inception_3a_5_5 = K.Conv2D(32,
                               5,
                               activation='relu',
                               name= 'inception_3a_5_5')(inception_3a_5_5_reduce)
    inception_3a_pool = K.MaxPooling2D(3,
                                    strides = 1,
                                    padding = 'same')(pool2_3_3)
    inception_3a_pool_1_1 = K.Conv2D(32,
                                    1,
                                    activation='relu',
                                    name='inception_3a_pool_1_1')(inception_3a_pool)
    
    # merge the inception_3a__
    inception_3a_output = K.Concatenate(axis=3)([inception_3a_1_1,
                                       inception_3a_3_3,
                                       inception_3a_5_5,
                                       inception_3a_pool_1_1])
    
    inception_3b_1_1 = K.Conv2D(128,
                              1,
                              activation='relu',
                              name = 'inception_3b_1_1')(inception_3a_output)
    inception_3b_3_3_reduce = K.Conv2D(128,
                                      1,
                                      activation = 'relu',
                                      name='inception_3b_3_3_reduce')(inception_3a_output)
    inception_3b_3_3 = K.Conv2D(192,
                               3,
                               activation = 'relu',
                               name='inception_3b_3_3')(inception_3b_3_3_reduce)
    inception_3b_5_5_reduce = K.Conv2D(32,
                                      1,
                                      activation='relu',
                                      name = 'inception_3b_5_5_reduce')(inception_3a_output)
    inception_3b_5_5 = K.Conv2D(96,
                               5,
                               name = 'inception_3b_5_5')(inception_3b_5_5_reduce)
    inception_3b_pool = K.MaxPooling2D(3,
                                     strides = 1,
                                     padding = 'same',
                                     name='inception_3b_pool')(inception_3a_output)
    inception_3b_pool_1_1 = K.Conv2D(64,
                                    kernel_size=1,
                                    activation='relu',
                                    name='inception_3b_pool_1_1')(inception_3b_pool)
    
    #merge the inception_3b_*
    inception_3b_output = K.Concatenate(axis=3,
                                      name='inception_3b_output')([inception_3b_1_1,
                                       inception_3b_3_3,
                                       inception_3b_5_5,
                                       inception_3b_pool_1_1])
    
    pool3_3_3 = K.MaxPooling2D(pool_size=3,
                             strides=2,
                             padding = 'same',
                             name='pool3_3_3')(inception_3b_output)
    inception_4a_1_1 = K.Conv2D(192,
                              kernel_size=1,
                              activation='relu',
                              name = 'inception_4a_1_1')(pool3_3_3)
    inception_4a_3_3_reduce = K.Conv2D(96,
                                     1,
                                     activation = 'relu',
                                     name = 'inception_4a_3_3_reduce')(pool3_3_3)
    inception_4a_3_3 = K.Conv2D(208,
                              3,
                              activation = 'relu',
                              name = 'inception_4a_3_3')(inception_4a_3_3_reduce)
    inception_4a_5_5_reduce = K.Conv2D(16,
                                     kernel_size=1,
                                     activation='relu',
                                     name='inception_4a_5_5_reduce')(pool3_3_3)
    inception_4a_5_5 = K.Conv2D(48,
                              kernel_size=5,
                              activation='relu',
                              name='inception_4a_5_5')(inception_4a_5_5_reduce)
    inception_4a_pool = K.MaxPooling2D(pool_size=3,
                                     strides=1,
                                     padding='same',
                                     name='inception_4a_pool')(pool3_3_3)
    inception_4a_pool_1_1 = K.Conv2D(64,
                                   kernel_size=1,
                                   activation='relu',
                                   name='inception_4a_pool_1_1')(inception_4a_pool)
    
    inception_4a_output = K.Concatenate(axis = 3,
                                      name = 'inception_4a_output')([inception_4a_1_1,
                                       inception_4a_3_3,
                                       inception_4a_5_5,
                                       inception_4a_pool_1_1])
    
    pool5_7_7 = K.AveragePooling2D(5,
                                   padding = 'same',
                                   strides = 1)(inception_4a_output)
    pool5_7_7 = K.Dropout(0.4)(pool5_7_7)
    output_layer = K.Dense(2,
                   activation='softmax')(pool5_7_7)
    model = Model(input_layer,output_layer)
    model.compile(optimizer=opt,
                  loss = loss_f)
    return model
```

**Comment by tobybreckon** [comment]
I am not very familiar with Keras directly.

Our example is heavily based on the standard inceptionV1 (googlenet) example, it is a cut down version of it as explained in the paper:
https://github.com/tflearn/tflearn/blob/master/examples/images/googlenet.py
in tflearn, which uses the same ops for concat.

Here is a Keras example of the same base inceptionV1 architecture:
https://gist.github.com/joelouismarino/a2ede9ab3928f999575423b9887abd14

I would suggest using this as a base. For example, in the Keras example for inceptionV1 they have ``axis=1``, whilst in tflearn ``axis=3`` used for concatenation to achieve the same effect it appears.
You have 3 in your code which is consistent with our tflearn code, but not the definition of inceptionV1 in Keras which uses ``axis=1``, try this perhaps.
